clang/lib/Driver/ToolChains/SYCL.cpp

```diff
@@ -27,6 +27,180 @@ using namespace clang::driver::tools;
+    {"arl_h", SYCLSupportedOffloadArchs::ARL_H},
+    {"bmg_g21", SYCLSupportedOffloadArchs::BMG_G21},
+    {"lnl_m", SYCLSupportedOffloadArchs::LNL_M},
+    // AMD GPU Mapping
+    {"gfx700", SYCLSupportedOffloadArchs::GFX700},
+    {"gfx701", SYCLSupportedOffloadArchs::GFX701},
+    {"gfx702", SYCLSupportedOffloadArchs::GFX702},
+    {"gfx801", SYCLSupportedOffloadArchs::GFX801},
+    {"gfx802", SYCLSupportedOffloadArchs::GFX802},
+    {"gfx803", SYCLSupportedOffloadArchs::GFX803},
+    {"gfx805", SYCLSupportedOffloadArchs::GFX805},
+    {"gfx810", SYCLSupportedOffloadArchs::GFX810},
+    {"gfx900", SYCLSupportedOffloadArchs::GFX900},
+    {"gfx902", SYCLSupportedOffloadArchs::GFX902},
+    {"gfx904", SYCLSupportedOffloadArchs::GFX904},
+    {"gfx906", SYCLSupportedOffloadArchs::GFX906},
+    {"gfx908", SYCLSupportedOffloadArchs::GFX908},
+    {"gfx909", SYCLSupportedOffloadArchs::GFX909},
+    {"gfx90a", SYCLSupportedOffloadArchs::GFX90A},
+    {"gfx90c", SYCLSupportedOffloadArchs::GFX90C},
+    {"gfx940", SYCLSupportedOffloadArchs::GFX940},
+    {"gfx941", SYCLSupportedOffloadArchs::GFX941},
+    {"gfx942", SYCLSupportedOffloadArchs::GFX942},
+    {"gfx1010", SYCLSupportedOffloadArchs::GFX1010},
+    {"gfx1011", SYCLSupportedOffloadArchs::GFX1011},
+    {"gfx1012", SYCLSupportedOffloadArchs::GFX1012},
+    {"gfx1013", SYCLSupportedOffloadArchs::GFX1013},
+    {"gfx1030", SYCLSupportedOffloadArchs::GFX1030},
+    {"gfx1031", SYCLSupportedOffloadArchs::GFX1031},
+    {"gfx1032", SYCLSupportedOffloadArchs::GFX1032},
+    {"gfx1033", SYCLSupportedOffloadArchs::GFX1033},
+    {"gfx1034", SYCLSupportedOffloadArchs::GFX1034},
+    {"gfx1035", SYCLSupportedOffloadArchs::GFX1035},
+    {"gfx1036", SYCLSupportedOffloadArchs::GFX1036},
+    {"gfx1100", SYCLSupportedOffloadArchs::GFX1100},
+    {"gfx1101", SYCLSupportedOffloadArchs::GFX1101},
+    {"gfx1102", SYCLSupportedOffloadArchs::GFX1102},
+    {"gfx1103", SYCLSupportedOffloadArchs::GFX1103},
+    {"gfx1150", SYCLSupportedOffloadArchs::GFX1150},
+    {"gfx1151", SYCLSupportedOffloadArchs::GFX1151},
+    {"gfx1200", SYCLSupportedOffloadArchs::GFX1200},
+    {"gfx1201", SYCLSupportedOffloadArchs::GFX1201},
+    // NVidia GPU Mapping.
+    {"sm_50", SYCLSupportedOffloadArchs::SM_50},
+    {"sm_52", SYCLSupportedOffloadArchs::SM_52},
+    {"sm_53", SYCLSupportedOffloadArchs::SM_53},
+    {"sm_60", SYCLSupportedOffloadArchs::SM_60},
+    {"sm_61", SYCLSupportedOffloadArchs::SM_61},
+    {"sm_62", SYCLSupportedOffloadArchs::SM_62},
+    {"sm_70", SYCLSupportedOffloadArchs::SM_70},
+    {"sm_72", SYCLSupportedOffloadArchs::SM_72},
+    {"sm_75", SYCLSupportedOffloadArchs::SM_75},
+    {"sm_80", SYCLSupportedOffloadArchs::SM_80},
+    {"sm_86", SYCLSupportedOffloadArchs::SM_86},
+    {"sm_87", SYCLSupportedOffloadArchs::SM_87},
+    {"sm_89", SYCLSupportedOffloadArchs::SM_89},
+    {"sm_90", SYCLSupportedOffloadArchs::SM_90},
+    {"sm_90a", SYCLSupportedOffloadArchs::SM_90A}};
```

Comment:
Can we, for AMD and Nvidia just reuse the existing `OffloadArchToStringMap` and `StringToOffloadArch` utility and the `OffloadArch` enum rather than having to duplicate the values again here? There is existing infrastructure here - https://github.com/intel/llvm/blob/sycl/clang/lib/Basic/Cuda.cpp#L85 already, and it will minimise the need to maintain the addition of these arches in both places. Also the Driver generally uses the existing `OffloadArch` enum accross the both for all offloading Cuda/Hip, OpenMP->Cuda/Hip. I'd be tempted to reuse these and make the current SYCLOffloadArchMap more of a IntelOffloadArchMap. Example use: https://github.com/intel/llvm/blob/sycl/clang/lib/Driver/Driver.cpp#L7652

We'd rather avoid code duplication. At some point it'll have to converge into one enum anyways so might as well reuse what's already there for Cuda/HIP.

Reply:
Hello,
I originally did re-use those existing enum/functions (you can check in my older commits), however, I decided to have a separate version specifically for SYCL for the following reasons.

1. According to the [SYCL Users-Manual](https://intel.github.io/llvm/UsersManual.html),  there is a specific list of supported targets for NVidia and AMD GPUs.

If you look at the [OffloadArch](https://github.com/intel/llvm/blob/sycl/clang/include/clang/Basic/Cuda.h#L55) enum, it includes AMD and NVidia GPU values that SYCL offloading does not support.
Furthermore, the way in which the AMD GPUs are listed in that enum, filtering the supported AMD GPU values becomes messy.

2. It made sense to me to contain all of SYCL offloading related stuff in SYCL.h rather than in Cuda.h/Cuda.cpp
I was even wondering if it was ok that OffloadArch which is in Cuda.h was containing AMD GPU targets in it.

Because of the above reasons I thought it would be better to have them separated. WDYT?

Reply:
> If you look at the [OffloadArch](https://github.com/intel/llvm/blob/sycl/clang/include/clang/Basic/Cuda.h#L55) enum, it includes AMD and NVidia GPU values that SYCL offloading does not support.
>
> 1. According to the [SYCL Users-Manual](https://intel.github.io/llvm/UsersManual.html), there is a specific list of supported targets for NVidia and AMD GPUs.
>
> If you look at the [OffloadArch](https://github.com/intel/llvm/blob/sycl/clang/include/clang/Basic/Cuda.h#L55) enum, it includes AMD and NVidia GPU values that SYCL offloading does not support.

This is okay, we don't need to support the full range, we will probably bump our minimum even higher at some point, at least to sm_52 to start with for Cuda (from currently sm_50) and likely for AMD soon as well. However, we should be able to reuse the same enum, maps and utility functions and just have reflect the [min;max] arch support range in `IsSYCLSupportedNvidiaGPUArch` helpers added to this PR.


> Furthermore, the way in which the AMD GPUs are listed in that enum, filtering the supported AMD GPU values becomes messy.

Understandable confusion. We have omitted three gfx archs for AMD that are supposed to be in the SYCL supported ones list or at least seem to have no good reason not to be there other than likely a genuine mistake, so I've just added them here [ [SYCL] Add missing supported AMDGPU architectures to SYCL #15723 ](https://github.com/intel/llvm/pull/15723). This means we should be able to do just:
for Nvidia
```
Arch >= OffloadArch::SM_50 && Arch <= OffloadArch::SM_90A
```
for AMD
```
Arch >= OffloadArch::GFX700 && Arch <= OffloadArch::GFX1201;
// or Arch >= OffloadArch::GFX700 && Arch < OffloadArch::AMDGCNSPIRV
```
This should be fine as we'll resolve the messy situation caused by us forgetting to add `GFX703`, `GFX704` and `GFX705` as supported.
I have no current knowledge of whether the recently added `GFX9/10/11/12/_GENERIC` targets would just work, but that should be easy to filter out inside `IsSYCLSupportedAmdGPUArch` for the time being.

> 2. It made sense to me to contain all of SYCL offloading related stuff in SYCL.h rather than in Cuda.h/Cuda.cpp
    I was even wondering if it was ok that OffloadArch which is in Cuda.h was containing AMD GPU targets in it.
I think you have a good point to be a little thrown of by all of this supposedly generic `OffloadArch` stuff to be part of `Cuda.h`. Historically, this was named `CudaArch` at first as Nvidia targets were the only ones added for offloading in the past with the upstream Cuda support. Later on AMD followed up with upstream support appending theirs and reusing and sharing a lot of existing Cuda offloading infrastructure for the HIP language. That being said, we should be able to move these common utilities wrt `OffloadArch` to a more common header because it is not really tied to Cuda per se.

> Because of the above reasons I thought it would be better to have them separated. WDYT?
I can think of 2 options here:
1. Reuse `OffloadArch` for Nvidia/AMD targets and have a `IntelOffloadArch` enum and associated utilities inside `SYCL.h` which will encapsulate the Intel CPU and GPU architectures supported for offloading SYCL.

2. Ideally, we'd push an upstream PR to `llvm-project` that moves the `OffloadArch` related utilities to a common header and implement them a counterpart `cpp` file moving the current implementations outside of `Cuda.cpp`.
See if upstream is okay with that change, and if they are we mimic it downstream here to avoid the wait for a pulldown and extend it by adding the Intel architectures there too. Otherwise, in case, upstream aren't okay or slow to come back, we do option 1.) that only reuses it for Nvidia/AMD and a separate `IntelOffloadArch` variant still lives inside `SYCL.h`.

The main goal is to avoid the duplication of the same utilities if there is really no good reason for it. If you do see more gaps in the architectures between the min and max we ought to support from the `OffloadArch` enum, do let me know and I'll take a look because there should be no difference for us (apart from those generic ones for AMD that I mentioned I am unsure about but we can easily just filter out in our helpers). Thanks!

Reply:
Thank you for the explanation and also for adding the missing GFX* entries.

> // plus potentially required filter out 'OffloadArch::GFX9/10/11/12/_GENERIC'

The GFX*_GENERIC values are sprinkled throughout the OffloadArch enum and this was one of my primary reasons for having a separate list for SYCL.

I'm wondering if there is a more elegant way of handling this filtering than how it is shown below:

```
(Arch >= OffloadArch::GFX700 && Arch < OffloadArch::GFX9_GENERIC) || (Arch >= OffloadArch::GFX900 && OffloadArch::GFX10_1_GENERIC) || (Arch >= OffloadArch::GFX1010 && Arch < OffloadArch::GFX10_3_GENERIC) || 
(Arch >= OffloadArch::GFX1030  && Arch < OffloadArch::GFX11_GENERIC) || (Arch >= OffloadArch::GFX1100 && Arch < OffloadArch::GFX12_GENERIC) || OffloadArch::GFX1200 || OffloadArch::GFX1201
```

I will go with Option 1 for now and reuse the existing utilities.

Reply:
Nice.
Well, I am not sure of any really elegant way to do that check but I feel like it will be fairly practical to do:

```cpp
Arch >= OffloadArch::GFX700 && Arch < OffloadArch::AMDGCNSPIRV && !IsGenericAMDGCNOffloadArch;
```

where we have the generic values checks in `bool IsGenericAMDGCNOffloadArch` so we can easily either remove the filter for generic gfx arches or extend it based on future support.

Note I put `Arch < OffloadArch::AMDGCNSPIRV` instead of `Arch <= OffloadArch::GFX1201` because it will be easier to keep up with AMD's active update of new gpu architectures support upstream. We don't currently list support for new ones but they'll most likely just work since AMD pushes all required compiler support upstream (maybe there could be a runtime part we may need to update sometimes but for the most part they'll work). Again, for now I am okay with this either way.

Thanks for all the replies and actions btw, that felt like a really a pleasant collaborative work!